### PR TITLE
[stable9] Use static method

### PR DESCRIPTION
**Pull request by LukasReschke**
This is ugly as hell. But if we directly access `\OCP\IUserSession::getUser` here PHP throws a segfault on some servers:

```
gdb /usr/sbin/apache2 --batch --quiet -ex "run -f /etc/apache2/apache2.conf -DNO_DETACH -DONE_PROCESS -DDEBUG $defines" -ex "quit"
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
AH00558: apache2: Could not reliably determine the server's fully qualified domain name, using 172.17.0.6. Set the 'ServerName' directive globally to suppress this message
warning: the debug information found in "/usr/lib/debug//usr/lib/php5/20121212/mysql.so" does not match "/usr/lib/php5/20121212/mysql.so" (CRC mismatch).

warning: the debug information found in "/usr/lib/debug/usr/lib/php5/20121212/mysql.so" does not match "/usr/lib/php5/20121212/mysql.so" (CRC mismatch).

warning: the debug information found in "/usr/lib/debug//usr/lib/php5/20121212/mysqli.so" does not match "/usr/lib/php5/20121212/mysqli.so" (CRC mismatch).

warning: the debug information found in "/usr/lib/debug/usr/lib/php5/20121212/mysqli.so" does not match "/usr/lib/php5/20121212/mysqli.so" (CRC mismatch).

warning: the debug information found in "/usr/lib/debug//usr/lib/php5/20121212/pdo_mysql.so" does not match "/usr/lib/php5/20121212/pdo_mysql.so" (CRC mismatch).

warning: the debug information found in "/usr/lib/debug/usr/lib/php5/20121212/pdo_mysql.so" does not match "/usr/lib/php5/20121212/pdo_mysql.so" (CRC mismatch).

[New Thread 0x7fffdd206700 (LWP 28386)]
[Thread 0x7fffdd206700 (LWP 28386) exited]
[New Thread 0x7fffdd206700 (LWP 28387)]
[New Thread 0x7fffdc975700 (LWP 28388)]

Program received signal SIGSEGV, Segmentation fault.
zend_parse_parameters (num_args=2, type_spec=type_spec@entry=0x7ffff3f85227 "s|s") at /build/php5-pO28mL/php5-5.5.9+dfsg/Zend/zend_API.c:923
923 /build/php5-pO28mL/php5-5.5.9+dfsg/Zend/zend_API.c: No such file or directory.
A debugging session is active.

    Inferior 1 [process 28382] will be killed.

Quit anyway? (y or n) [answered Y; input not from terminal]
```

I'm still struggling to get a minimal example of this (also considering that it doesn't happen on every system :tada: ). Anyways, it's a functional workaround until I determined the exact bug… (also given the fact that patches may never be backported to distributions :see_no_evil:

Fixes https://github.com/owncloud/user_shibboleth/issues/128

cc @karlitschek @cmonteroluque @blizzz @davitol @rperezb

### CODE REVIEW

**Comment by blizzz** [comment]
@heiglandreas may I ask you whether you might have an  idea what is happening here on PHP side?

**Comment by heiglandreas** [comment]
Currently swamped in work and jumping into the next meeting. I'll have a look at it this evening!

**Comment by davitol** [comment]
:+1: WFM :D

**Comment by DeepDiver1975** [comment]
Merge? @cmonteroluque

**Comment by heiglandreas** [comment]
From a debugging POV the interesting question is where *exactly* the SEGFAULT happens. The gdb-output is missing the most interesting part as to where the segfault actually happens *in the PHP-code*. The ```zend_parse_parameter``` only tells me that a function call is parsed that takes at least one string-parameter and all other parameters are parsed as string as well. But which PHP-side function call that is, I can't say.

The other question is, where exactly that behaviour happens. Is it only on a certain device-type? As PHP5.5.9 is not a current PHP-Version I believe we're talking about Ubuntus 14.04 LTS systems that have the issue. But does that affect all systems? Under which circumstances? Is there a test that reliably triggers that SEGFAULT?

Apart from that the patch looks OK from my sied but I have no experience with OC (apart from running one on my own server :wink:). I've added a comment though :wink:

File: lib/private/log/owncloud.php [hunk 1]
```diff
@@ -91,11 +91,10 @@ public static function write($app, $message, $level) {
 		$url = ($request->getRequestUri() !== '') ? $request->getRequestUri() : '--';
 		$method = is_string($request->getMethod()) ? $request->getMethod() : '--';
 		if(\OC::$server->getConfig()->getSystemValue('installed', false)) {
-			$userObj = \OC::$server->getUserSession()->getUser();
+			$user = (\OC_User::getUser()) ? \OC_User::getUser() : '--';
 		} else {
-			$userObj = null;
+			$user = '--';
```
Comment (heiglandreas) [inline lib/private/log/owncloud.php]: I'd put that as declaration of the variable before the ```if``` in line 93 and would then get rid of the complete ```else```-block.

**Comment by LukasReschke** [comment]
Thanks @heiglandreas. Some more information below if you find some spare minutes :rocket: 

> From a debugging POV the interesting question is where exactly the SEGFAULT happens. The gdb-output is missing the most interesting part as to where the segfault actually happens in the PHP-code. The zend_parse_parameter only tells me that a function call is parsed that takes at least one string-parameter and all other parameters are parsed as string as well. But which PHP-side function call that is, I can't say.

More complete trace (PHP 5.6.20-1+deb.sury.org~trusty+1) can be found at https://gist.github.com/LukasReschke/230c4778791e910d0451b58eeda56c09. Some excerpts:

```
#0  malloc_printerr (ptr=<optimized out>, str=<optimized out>, action=<optimized out>) at malloc.c:4992
        buf = "0\335\343VUU\000\000\360\317?Ww\002\000\000\300"
        cp = <optimized out>
#1  malloc_consolidate (av=av@entry=0x7ffff751e760 <main_arena>) at malloc.c:4157
        fb = <optimized out>
        maxfb = <optimized out>
        p = <optimized out>
        nextp = <optimized out>
        unsorted_bin = <optimized out>
        first_unsorted = <optimized out>
        nextchunk = <optimized out>
        size = <optimized out>
        nextsize = <optimized out>
        prevsize = <optimized out>
        nextinuse = <optimized out>
        bck = <optimized out>
        fwd = <optimized out>
        __func__ = "malloc_consolidate"
#2  0x00007ffff71dfdf8 in _int_malloc (av=0x7ffff751e760 <main_arena>, bytes=4060) at malloc.c:3423
        nb = 4080
        idx = 98
        bin = <optimized out>
        victim = <optimized out>
        size = <optimized out>
        victim_index = <optimized out>
        remainder = <optimized out>
        remainder_size = <optimized out>
        block = <optimized out>
        bit = <optimized out>
        map = <optimized out>
        fwd = <optimized out>
        bck = <optimized out>
        errstr = 0x0
        __func__ = "_int_malloc"
#3  0x00007ffff71e27b0 in __GI___libc_malloc (bytes=4060) at malloc.c:2891
        ar_ptr = 0x7ffff751e760 <main_arena>
        victim = 0xff0
        __func__ = "__libc_malloc"
#4  0x00007fffed1ed565 in ber_memalloc_x () from /usr/lib/x86_64-linux-gnu/liblber-2.4.so.2
No symbol table info available.
#5  0x00007fffed1ebbfb in ber_realloc () from /usr/lib/x86_64-linux-gnu/liblber-2.4.so.2
No symbol table info available.
#6  0x00007fffed1eab3b in ?? () from /usr/lib/x86_64-linux-gnu/liblber-2.4.so.2
No symbol table info available.
#7  0x00007fffed1eb684 in ber_printf () from /usr/lib/x86_64-linux-gnu/liblber-2.4.so.2
No symbol table info available.
#8  0x00007fffe5ec103f in zif_ldap_control_paged_result (ht=4, return_value=0x5555573e37d8, return_value_ptr=<optimized out>, 
    this_ptr=<optimized out>, return_value_used=<optimized out>) at /build/php5-6hC5Fj/php5-5.6.20+dfsg/ext/ldap/ldap.c:2811
        pagesize = 0
        iscritical = 0 '\000'
        link = 0x55555637d210
        cookie = 0x7fffdc0b6400 ""
        cookie_len = 0
        lcookie = {bv_len = 0, bv_val = 0x7fffdc0b6400 ""}
        ld = <optimized out>
        ldap = 0x5555560cfeb0
        ber = 0x555555a8bd70
        ctrl = {ldctl_oid = 0x200000019 <error: Cannot access memory at address 0x200000019>, ldctl_value = {bv_len = 0, bv_val = 0x0}, 
          ldctl_iscritical = 0 '\000'}
        ctrlsp = {0x0, 0x0}
        rc = <optimized out>
        myargcount = 4
#9  0x00007ffff3eca8fb in dtrace_execute_internal (execute_data_ptr=<optimized out>, fci=<optimized out>, return_value_used=<optimized out>)
    at /build/php5-6hC5Fj/php5-5.6.20+dfsg/Zend/zend_dtrace.c:97
        lineno = <optimized out>
        filename = <optimized out>
#10 0x00007ffff3ecc605 in zend_call_function (fci=fci@entry=0x7fffff7ff510, fci_cache=fci_cache@entry=0x7fffff7ff4e0)
    at /build/php5-6hC5Fj/php5-5.6.20+dfsg/Zend/zend_execute_API.c:849
        call_via_handler = 0
        i = <optimized out>
        original_return_value = <optimized out>
        calling_symbol_table = <optimized out>
        original_op_array = <optimized out>
        original_opline_ptr = <optimized out>
        current_scope = 0x5555561b6350
        current_called_scope = 0x5555561b6350
        calling_scope = <optimized out>
        called_scope = 0x0
        current_this = 0x5555561b29b0
        execute_data = {opline = 0x0, function_state = {function = 0x555555934150, arguments = 0x5555573fcf68}, op_array = 0x0, object = 0x0, 
          symbol_table = 0x0, prev_execute_data = 0x5555573fce58, old_error_reporting = 0x0, nested = 0 '\000', 
          original_return_value = 0x5555573fcce8, current_scope = 0x5555561b6350, current_called_scope = 0x5555561b6350, 
          current_this = 0x5555561b29b0, fast_ret = 0x0, delayed_exception = 0x0, call_slots = 0x5555573fcf10, call = 0x5555573fcf10}
        fci_cache_local = {initialized = 252 '\374', function_handler = 0x7ffff3ee5358 <zend_parse_parameters+152>, 
          calling_scope = 0x7ffff466b6fc <php_execute.return_semaphore>, called_scope = 0x3000000028, object_ptr = 0x7fffff7ff4d0}
```

> The other question is, where exactly that behaviour happens. Is it only on a certain device-type? As PHP5.5.9 is not a current PHP-Version I believe we're talking about Ubuntus 14.04 LTS systems that have the issue. 

Correct. I installed 5.6.20-1+deb.sury.org~trusty+1 as well on it but it's failing famously as well as seen above.

> But does that affect all systems?

I can reproduce on one of test systems that we have here with. But it works fine on others. Seems to indicate some LDAP problem? :confused: 

@blizzz FYI - I have disabled Shibboleth on the system and just tried to login with an LDAP account. Explodes there as well :see_no_evil: 

@owncloud/qa Can you provide more information about the LDAP system that the docker instance on 51924 uses? - Any chance that we can get the Docker files of all of the systems in case so that we can reproduce this on another system?

>  Under which circumstances? 

:see_no_evil: :hear_no_evil: :speak_no_evil: - No proper idea until now.

> Is there a test that reliably triggers that SEGFAULT?

I didn't yet manage to get this down into a minimum example down.

**Comment by LukasReschke** [comment]
@owncloud/qa Also is there anything in the LDAP server logs?

**Comment by heiglandreas** [comment]
Thanks for the update @LukasReschke! I'll dig through it :) - Any hints on what LDAP-lib the php is linked against would be highly appreciated! ;)

**Comment by heiglandreas** [comment]
Thanks for the update @LukasReschke! I'll dig through it :) - Any hints on what LDAP-lib the php is linked against would be highly appreciated! ;)

**Comment by LukasReschke** [comment]
@heiglandreas phpinfo claims for the LDAP stuff:

```
RCS Version	$Id: 5ffc17b96f6abc34bc3607bc1131ea7a79586613 $
Total Links	0/unlimited
API Version	3001
Vendor Name	OpenLDAP
Vendor Version	20431
SASL Support	Enabled
```

**Comment by MorrisJobke** [comment]
@LukasReschke Port to master?

**Comment by LukasReschke** [comment]
I'd prefer to get a more proper analysis first before cherry-picking that into master. Let's give us some more time here.

**Comment by heiglandreas** [comment]
I need a bit more time here to analyze that. The logfiles that @LukasReschke requested would also be handy :wink:

But I'll dig through this anyway!

**Comment by LukasReschke** [comment]
I'll try to get access to that machine…

cc @jmaciasportela @rperezb @cmonteroluque

**Comment by jmaciasportela** [comment]
You have access now @LukasReschke

**Comment by heiglandreas** [comment]
To be honest, I find it difficult to get the connection between [/lib/private/user/session.php:148](https://github.com/owncloud/core/blob/682821c71ed2af297170294632931be1ac769f25/lib/private/user/session.php#L148) and [/apps/user_ldap/lib/access.php#884](https://github.com/owncloud/core/blob/fe2a506e56e33625edb51e6b5e0df8911cf4e294/apps/user_ldap/lib/access.php#L884) which is the only occurrence of the ```ldap_control_paged_result```. 

How does a request for logging cause a paged search in LDAP (meaning there are expected more than 100 results from a search)? Would it be possible to get a stack-trace for that call to ```\OC::$server->getUserSession()->getUser()```? I would need to setup the whole process and it might be easier for you to just fire up the process in question :wink:  

Thanks for your input!

**Comment by blizzz** [comment]
> How does a request for logging cause a paged search in LDAP (meaning there are expected more than 100 results from a search)?

We go directly to paged search, as it appeared that mixing paged search and normal one cause conflicts. However, this was ages ago and might have been due to wrong cancellation of paged searches, I did not revisit this since. So, of course for login there is ideally just one response expected, it would not be really necessary here.

**Comment by blizzz** [comment]
> Would it be possible to get a stack-trace for that call to \OC::$server->getUserSession()->getUser()?

@LukasReschke do you still have the setup for it?

**Comment by LukasReschke** [comment]
Yes. Will try to get a stacktrace, let me see how I can do that :)

**Comment by heiglandreas** [comment]
When I need one I always include something like this:

```php
try {
   throw new \Exception('Whops');
} catch (\Exception $e) {
   var_Dump($e->getTraceAsString());
}
```

Might not be the most elegant solution, but it does the job :wink:

**Comment by MorrisJobke** [comment]
> Might not be the most elegant solution, but it does the job

```php
\OC::$server->getLogger()->logException(new \Exception());
```

😉

**Pull request merged**